### PR TITLE
LwDetrImageLoss: Fix dtype casting to prevent crash when using amp on cuda device

### PR DESCRIPTION
**Pull request by m-matthias**
# What does this PR do?

Prevent crash in class LwDetrImageLoss when using it with float16 automatic mixed precision on a Cuda device.
torch.pow causes an autocast to float32 when used with Cuda, which caused a type mismatch at ``` pos_weights[pos_ind] = t ```.
This is fixed by casting back to the original type after the torch.pow operations.
See also [torch.amp](https://docs.pytorch.org/docs/stable/amp.html#cuda-ops-that-can-autocast-to-float32)
Also added ``` pos_ind = tuple(pos_ind) ``` to fix a deprecation warning.

Fixes [44857](https://github.com/huggingface/transformers/issues/44857)

## Who can review?
@sbucaille 
@yonigozlan

### CODE REVIEW

File: src/transformers/loss/loss_lw_detr.py [hunk 1]
```diff
@@ -128,13 +128,14 @@ def loss_labels(self, outputs, targets, indices, num_boxes):
         prob = source_logits.sigmoid()
         # init positive weights and negative weights
         pos_weights = torch.zeros_like(source_logits)
-        neg_weights = prob**gamma
+        neg_weights = prob.pow(gamma).to(source_logits.dtype)
 
         pos_ind = list(idx)
         pos_ind.append(target_classes_o)
+        pos_ind = tuple(pos_ind)
```
Comment (sbucaille) [inline src/transformers/loss/loss_lw_detr.py]: Since `idx` is already a tuple we can just use 
```
pos_ind = idx + (target_classes_o,)
```
instead of converting to a list then converting back to a tuple, 3 lines in 1

File: src/transformers/loss/loss_lw_detr.py [hunk 2]
```diff
@@ -128,13 +128,14 @@ def loss_labels(self, outputs, targets, indices, num_boxes):
         prob = source_logits.sigmoid()
         # init positive weights and negative weights
         pos_weights = torch.zeros_like(source_logits)
-        neg_weights = prob**gamma
+        neg_weights = prob.pow(gamma).to(source_logits.dtype)
```
Comment (sbucaille) [inline src/transformers/loss/loss_lw_detr.py]: Since we end up with 3 `source_logits.dtype`, we can declare `dtype` below like that : 
```
source_logits = outputs["logits"]
dtype = source_logits.dtype
```
and refer to dtype on occurences

**Review by sbucaille — COMMENTED** [review]
Thanks for the fix !

File: src/transformers/loss/loss_lw_detr.py (hunk 1 above)
Comment (m-matthias) [reply src/transformers/loss/loss_lw_detr.py]: done

File: src/transformers/loss/loss_lw_detr.py (hunk 2 above)
Comment (m-matthias) [reply src/transformers/loss/loss_lw_detr.py]: done

**Comment by m-matthias** [comment]
@sbucaille Could you maybe have a short look at the tests please? I dont really understand why some of them are failing.

**Comment by sbucaille** [comment]
@m-matthias don't worry it's unrelated to the PR

**Review by yonigozlan — APPROVED** [review]
Thanks for fixing @m-matthias and for reviewing @sbucaille ! I added a short comment to explain the changes, and apply a similar fix to rt_detr. LGTM!
